query.ml

```diff
@@ -170,13 +229,30 @@ let infer' c0 query =
     | Var var -> tuck constraints (On_var (var, Eq_any))
     | Strings (`Var var) -> tuck constraints (On_var (var, Eq_list String))
     | Strings (`List l) -> l |> List.iter (function var -> tuck constraints (On_var (var, Eq_type String)))
+    | Nested { path=_; inner_query; inner_hits } ->
+      iter inner_query;
+      begin match inner_hits with
+      | None -> ()
+      | Some ih ->
+        (match ih.size with Some (`Var v) -> tuck constraints (On_var (v, Eq_type Int)) | _ -> ());
+        (match ih.from with Some (`Var v) -> tuck constraints (On_var (v, Eq_type Int)) | _ -> ());
+      end
     | Nothing -> ()
   in
   iter query;
   !constraints
 
 let infer = infer' []
 
+let rec extract_inner_hits_from_query query =
+  match query.query with
+  | Nested { path; inner_hits = Some spec; _ } ->
+    [path, spec]
+  | Nested { inner_hits = None; _ } -> []
+  | Bool clauses ->
+    clauses |> List.map snd |> List.flatten |> List.map extract_inner_hits_from_query |> List.flatten
+  | _ -> []
```

Comment:
I think there aren't too many constructors in this variant, would be nice not to use a wildcard here